src/test/kotlin/com/pubnub/api/legacy/endpoints/pubsub/SignalTest.kt

```diff
@@ -204,4 +212,87 @@ class SignalTest : BaseTest() {
     }
+
+    @Test
+    fun testSpaceIdQueryParamIsPassed() {
+        val spaceIdValue = "thisIsSpaceId"
+
+        stubFor(
+            get(WireMock.urlPathEqualTo("/signal/myPublishKey/mySubscribeKey/0/$channel/0/%22hi%22"))
+                .willReturn(aResponse().withBody("""[1,"Sent","15883272000000000"]"""))
+        )
+
+        pubnub.signal(
+            channel = channel,
+            message = "hi",
+            spaceId = SpaceId(spaceIdValue)
+        ).sync()!!
+
+        val requests = findAll(getRequestedFor(urlMatching("/.*")))
+        assertEquals(1, requests.size)
+        MatcherAssert.assertThat(
+            requests[0].queryParams.map { (k, v) -> k to v.firstValue() }.toMap(),
+            Matchers.hasEntry(Publish.SPACE_ID_QUERY_PARAM, spaceIdValue)
+        )
+    }
+
+    @Test
+    fun testMissingSpaceIdQueryParamIsNotSet() {
+        stubFor(
+            get(WireMock.urlPathEqualTo("/signal/myPublishKey/mySubscribeKey/0/$channel/0/%22hi%22"))
+                .willReturn(aResponse().withBody("""[1,"Sent","15883272000000000"]"""))
+        )
+
+        pubnub.signal(
+            channel = channel,
+            message = "hi"
+        ).sync()!!
+
+        val requests = findAll(getRequestedFor(urlMatching("/.*")))
+        assertEquals(1, requests.size)
+        MatcherAssert.assertThat(requests[0].queryParams, Matchers.not(Matchers.hasKey(Publish.SPACE_ID_QUERY_PARAM)))
+    }
+
+    @Test
+    fun testMessageTypeQueryParamIsPassedInPublish() {
+        val messageTypeValue = "thisIsSpaceId"
+
+        stubFor(
+            get(WireMock.urlPathEqualTo("/signal/myPublishKey/mySubscribeKey/0/$channel/0/%22hi%22"))
+                .willReturn(aResponse().withBody("""[1,"Sent","15883272000000000"]"""))
+        )
+
+        pubnub.signal(
+            channel = channel,
+            message = "hi",
+            messageType = MessageType(messageTypeValue)
+        ).sync()!!
+
+        val requests = findAll(getRequestedFor(urlMatching("/.*")))
+        assertEquals(1, requests.size)
+        MatcherAssert.assertThat(
+            requests[0].queryParams.map { (k, v) -> k to v.firstValue() }.toMap(),
+            Matchers.hasEntry(Publish.MESSAGE_TYPE_QUERY_PARAM, messageTypeValue)
+        )
+    }
+
+    @Test
+    fun testMissingMessageTypeQueryParamIsNotSet() {
+        stubFor(
+            get(WireMock.urlPathEqualTo("/signal/myPublishKey/mySubscribeKey/0/$channel/0/%22hi%22"))
+                .willReturn(aResponse().withBody("""[1,"Sent","15883272000000000"]"""))
+        )
+
+        pubnub.signal(
+            channel = channel,
+            message = "hi"
+        ).sync()!!
+
+        val requests = findAll(getRequestedFor(urlMatching("/.*")))
+        assertEquals(1, requests.size)
+        MatcherAssert.assertThat(
+            requests[0].queryParams,
+            Matchers.not(Matchers.hasKey(Publish.MESSAGE_TYPE_QUERY_PARAM))
+        )
+    }
```

Comment:
PR looks good to me. One think that I would add is unit test that covers changes in SubcribeMessageProcessor, something similar as in Java:

  @Test
    public void pnSignalResult_should_contain_messageType_and_spaceId_when_they_are_provided_in_subscribeMessage() throws PubNubException {
        //given
        String message = "signal content something interesting";
        String messageType = "myMessageType";
        String spaceId = "mySpace";
        String subscribeResponseJson = getSubscribeResponseForSignalWithMessageTypeAndSpaceId(message, messageType, spaceId);

        Gson gson = new Gson();
        SubscribeMessageProcessor subscribeMessageProcessor = subscribeMessageProcessor(config());
        SubscribeEnvelope subscribeEnvelope = gson.fromJson(subscribeResponseJson, SubscribeEnvelope.class);

        //when
        PNEvent result = subscribeMessageProcessor.processIncomingPayload(subscribeEnvelope.getMessages().get(0));

        //then
        PNSignalResult signalResult = (PNSignalResult) result;

        assertThat(result, is(instanceOf(PNSignalResult.class)));
        assertThat(signalResult.getMessage().getAsString(), is(message));
        assertThat(signalResult.getMessageType().getValue(), equalTo(messageType));
        assertThat(signalResult.getSpaceId().getValue(), equalTo(spaceId));


    }

    private String getSubscribeResponseForSignalWithMessageTypeAndSpaceId(String message, String messageType, String spaceId) {
        return "{\n" +
                "  \"t\": {\n" +
                "    \"t\": \"16734390751423544\",\n" +
                "    \"r\": 41\n" +
                "  },\n" +
                "  \"m\": [\n" +
                "    {\n" +
                "      \"a\": \"1\",\n" +
                "      \"f\": 0,\n" +
                "      \"e\": 1,\n" +
                "      \"i\": \"client-a9e92988-4909-421a-a66e-96c778bc9d86\",\n" +
                "      \"p\": {\n" +
                "        \"t\": \"16734390751423544\",\n" +
                "        \"r\": 41\n" +
                "      },\n" +
                "      \"k\": \"sub-c-cb8b98b4-cd27-11ec-b360-1a35c262c233\",\n" +
                "      \"c\": \"ch_mqwvjygftw\",\n" +
                "     \"d\":  \"" + message + "\",\n" +
                "      \"mt\": \"" + messageType + "\",\n" +
                "      \"si\": \"" + spaceId + "\"\n" +
                "    }\n" +
                "  ]\n" +
                "}";
    }

Reply:
Done